R/remotes.R

```diff
@@ -394,6 +473,110 @@ renv_remotes_resolve_github <- function(entry) {
 
 }
 
+renv_remotes_resolve_git <- function(parsed) {
+
+  # handle refs that are direct refs
+  # handle potential PR refs
+  # handle tags too?
+
+
+  # For now, set package & version to dummy values.
+  # will need to clone first to resolve these.
+  temp_package <- parsed$repo
+  temp_version <- "UNKNOWN"
+  repo         <- parsed$repo
+  url          <- parsed$url
+  subdir       <- parsed$subdir
+
+  # handle git ref
+  pull <- parsed$pull %||% ""
+  ref <- parsed$ref %||% ""
+
+  remote_ref <- case(
+    nzchar(pull) ~ renv_remotes_resolve_git_pull(pull),
+    nzchar(ref)  ~ ref,
+    TRUE         ~ ""
+  )
+
+  record <- list(
+    Package        = temp_package,
+    Version        = temp_version,
+    Source         = "git",
+    RemoteType     = "git",
+    RemoteUrl      = url,
+    RemoteSubdir   = subdir,
+    RemoteRef      = remote_ref
+  )
+
+  desc <- renv_remotes_resolve_git_description(record)
+
+  record$Package <- desc$Package
+  record$Version <- desc$Version
+
+  record
+}
+
+
+renv_remotes_resolve_git_sha_ref <- function(record) {
+  renv_git_preflight()
+  template <- c(
+    "git ls-remote \"${ORIGIN}\" \"${REF}\" | head -1 | sed 's/\t.*//'"
+  )
+
+  data <- list(
+    ORIGIN = record$RemoteUrl,
+    REF    = record$RemoteRef %||% record$RemoteSha
+  )
+
+  commands <- renv_template_replace(template, data)
+  command <- paste(commands, collapse = " && ")
+  if (renv_platform_windows())
+    command <- paste(comspec(), "/C", command)
+
+  sha <- local({
+    renv_scope_auth(record)
+    renv_scope_git_auth()
+    system(command, intern = TRUE)
+  })
+
+  command_failed <- attr(sha, "status") %||% 0
+
+  if (command_failed != 0L) {
+    fmt <- "Checking %s for changes: `'%s'` failed  [status code %i]"
+    stopf(fmt, record$Package, command, status)
+  }
+
+  return(sha)
+}
+
+
+renv_remotes_resolve_git_description <- function(record) {
+  path <- tempfile("renv-git-")
+  ensure_directory(path)
+
+  renv_retrieve_git_impl(record, path)
```

Comment:
Do you know if it is possible to check out only a single file from a git repository? It would be nice to be able to avoid a whole clone just to get the DESCRIPTION.

Reply:
>  It would be nice to be able to avoid a whole clone just to get the DESCRIPTION.

It sure would- unfortunately, from googling around on the subject, there doesn't seem to be many robust, portable options for this.

- Some of the git mechanisms require the following not-so-portable conditions:
  - the remote git server has specific options enabled, which some (eg github) don't
  - git client version >= 2.25ish (that's fairly modern & not likely on many distros, though `renv` does already require a minimum git version that's somewhat modern)
- OR, they require the refspec meet restrictions that don't work for us, such as being an actual formal ref, as opposed to a commit SHA, etc.

Future work might be to write up an implementation that can try out some of these options to reduce overhead, however, the script would probably be non-trivial, and involve stuff like:

- try out mechanism 1, check the return code
- try out mechanism 2, check it was a valid ref that could be fetched
- etc

Because the `git` protocol should be as general as possible to any `git` repo, unfortunately I think this is still the best bet, and we can try to improve upon this down the line.

Reply:
I'll leave this unresolved so you can have a read of it, but for now, I think we should accept the "clone whole repo" implementation as "good enough"

Reply:
Thanks! Given your investigation I agree. (Perhaps in the future we could implement something like "if git.version >= 2.25, then use the clever implementation; otherwise, fall back to the portable one")